streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```diff
@@ -73,203 +57,38 @@ public static boolean isValidCleanupPolicy(final String cleanupPolicy) {
         return true;
     }
 
-    private class ZKStringSerializer implements ZkSerializer {
-
-        /**
-         * @throws AssertionError if the byte String encoding type is not supported
-         */
-        @Override
-        public byte[] serialize(Object data) {
-            try {
-                return ((String) data).getBytes("UTF-8");
-            } catch (UnsupportedEncodingException e) {
-                throw new AssertionError(e);
-            }
-        }
-
-        /**
-         * @throws AssertionError if the byte String encoding type is not supported
-         */
-        @Override
-        public Object deserialize(byte[] bytes) {
-            try {
-                if (bytes == null)
-                    return null;
-                else
-                    return new String(bytes, "UTF-8");
-            } catch (UnsupportedEncodingException e) {
-                throw new AssertionError(e);
-            }
-        }
-    }
-
     public InternalTopicManager() {
-        this.zkClient = null;
         this.replicationFactor = 0;
         this.windowChangeLogAdditionalRetention = WINDOW_CHANGE_LOG_ADDITIONAL_RETENTION_DEFAULT;
     }
 
-    public InternalTopicManager(String zkConnect, final int replicationFactor, long windowChangeLogAdditionalRetention) {
-        this.zkClient = new ZkClient(zkConnect, 30 * 1000, 30 * 1000, new ZKStringSerializer());
+    public InternalTopicManager(StreamsConfig config, final int replicationFactor, long windowChangeLogAdditionalRetention) {
+        this.config = config;
+        this.streamsKafkaClient = new StreamsKafkaClient(config);
         this.replicationFactor = replicationFactor;
         this.windowChangeLogAdditionalRetention = windowChangeLogAdditionalRetention;
     }
 
+    /**
+     * If the topic exists do nothing (we don't change partititions or delete the existing topics).
+     * Otherwise create the new topic.
+     * @param topic
+     * @param numPartitions
+     */
     public void makeReady(InternalTopicConfig topic, int numPartitions) {
-        boolean topicNotReady = true;
 
-        while (topicNotReady) {
-            Map<Integer, List<Integer>> topicMetadata = getTopicMetadata(topic.name());
-
-            if (topicMetadata == null) {
-                try {
-                    createTopic(topic, numPartitions, replicationFactor);
-                } catch (ZkNodeExistsException e) {
-                    // ignore and continue
+        if (!streamsKafkaClient.topicExists(topic.name())) {
+            streamsKafkaClient.createTopic(topic, numPartitions, replicationFactor, windowChangeLogAdditionalRetention);
+        } else {
+            MetadataResponse.TopicMetadata topicMetadata = streamsKafkaClient.getTopicMetadata(topic.name());
```

Comment:
I'm thinking in `StreamsKafkaClient.getTopicMetadata(..)` we could just return the `MetadataResponse.TopicMetadata` regardless of the `error().code()`.  So then i think we will never get `null` here as i believe we will always get a response. We can then throw an exception with a more meaningful  message, i.e., based on the `error().code()` - rather than just "Topic metadata is corrupted"